// @morpheus/cli/dist/definition.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OS = exports.ANDROID_CPU = exports.EDUCATION_LICENSE = exports.EDU_LICENSE_DIR = exports.LICENSE_DIR = exports.LICENSE_FILE_NAME = exports.EDUCATION_APP_ID = exports.PACKAGE_GROUP_NAME = void 0;
/* eslint-disable camelcase */
exports.PACKAGE_GROUP_NAME = '@morpheus';
exports.EDUCATION_APP_ID = 'mcore.edu.*';
exports.LICENSE_FILE_NAME = 'mcore.mobile.lic';
exports.LICENSE_DIR = 'license';
exports.EDU_LICENSE_DIR = 'edu_license';
exports.EDUCATION_LICENSE = {
    appId: exports.EDUCATION_APP_ID,
    bundleId: exports.EDUCATION_APP_ID,
    packageName: exports.EDUCATION_APP_ID,
    expirationDate: '9999-12-31',
};
var ANDROID_CPU;
(function (ANDROID_CPU) {
    ANDROID_CPU["ARM64_V8A"] = "arm64-v8a";
    ANDROID_CPU["ARMEABI_V7A"] = "armeabi-v7a";
    ANDROID_CPU["ARMEABI"] = "armeabi";
    ANDROID_CPU["X86"] = "x86";
    ANDROID_CPU["X86_64"] = "x86_64";
})(ANDROID_CPU = exports.ANDROID_CPU || (exports.ANDROID_CPU = {}));
var OS;
(function (OS) {
    OS["Unknown"] = "unknown";
    OS["Mac"] = "mac";
    OS["Windows"] = "windows";
    OS["Linux"] = "linux";
})(OS = exports.OS || (exports.OS = {}));
