// @morpheus/cli/dist/guards.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isAndroidAppIcon = exports.isExtConfig = exports.isLicense = exports.isMorpheusPackage = exports.isNpmAsset = exports.isAndroidCpu = exports.isSuperAgentError = void 0;
const definition_1 = require("./definition");
function isSuperAgentError(err) {
    return err && err.response && typeof err.response === 'object';
}
exports.isSuperAgentError = isSuperAgentError;
function isAndroidCpu(value) {
    const cpus = Object.values(definition_1.ANDROID_CPU);
    return value && typeof value === 'string' && cpus.includes(value);
}
exports.isAndroidCpu = isAndroidCpu;
function isNpmAsset(value) {
    return value && typeof value.name === 'string' && typeof value.version === 'string';
}
exports.isNpmAsset = isNpmAsset;
function isMorpheusPackage(value) {
    return value && value.morpheus;
}
exports.isMorpheusPackage = isMorpheusPackage;
function isLicense(value) {
    return (value &&
        typeof value.appId === 'string' &&
        typeof value.packageName === 'string' &&
        typeof value.bundleId === 'string' &&
        typeof value.expirationDate === 'string');
}
exports.isLicense = isLicense;
function isExtConfig(value) {
    return (value &&
        typeof value.applicationId === 'string' &&
        typeof value.androidPackageName === 'string' &&
        typeof value.iosBundleId === 'string');
}
exports.isExtConfig = isExtConfig;
function isAndroidAppIcon(value) {
    return value && typeof value.density === 'string' && typeof value.imgPath === 'string';
}
exports.isAndroidAppIcon = isAndroidAppIcon;
